### PR TITLE
8307185: pkcs11 native libraries make JNI calls into java code while holding GC lock

### PR DESCRIPTION
**Pull request by djelinski**
This patch fixes random deadlocks in PKCS11 decryption and encryption code.

The deadlocks were caused by object allocation in `ckAssertReturnValueOK` waiting for GC; GC was in turn waiting for `ReleasePrimitiveArrayCritical`, which never happened.

The fix moves the call to `ckAssertReturnValueOK` after `ReleasePrimitiveArrayCritical`.

All tests in sun/security/pkcs11 were executed with `JTREG=JAVA_OPTIONS=-Xcheck:jni` with and without this patch.
Without this patch, 5 tests produced the following warning:
```
Warning: Calling other JNI functions in the scope of Get/ReleasePrimitiveArrayCritical or Get/ReleaseStringCritical
```
With this patch the warning was not observed.

Tier2-3 tests are still green.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8307185](https://bugs.openjdk.org/browse/JDK-8307185): pkcs11 native libraries make JNI calls into java code while holding GC lock (**Bug** - P2)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [64f4894a](https://git.openjdk.org/jdk/pull/14931/files/64f4894a764a3dbec6a654711e411cdde425696a)
 * [Valerie Peng](https://openjdk.org/census#valeriep) (@valeriepeng - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14931/head:pull/14931` \
`$ git checkout pull/14931`

Update a local copy of the PR: \
`$ git checkout pull/14931` \
`$ git pull https://git.openjdk.org/jdk.git pull/14931/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14931`

View PR using the GUI difftool: \
`$ git pr show -t 14931`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14931.diff">https://git.openjdk.org/jdk/pull/14931.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14931#issuecomment-1642004002)

### CODE REVIEW

**Review by AlanBateman — APPROVED** [review]

File: src/jdk.crypto.cryptoki/share/native/libj2pkcs11/p11_crypt.c [hunk 1]
```diff
@@ -166,15 +169,13 @@ Java_sun_security_pkcs11_wrapper_PKCS11_C_1Encrypt
                                     (CK_BYTE_PTR)(outBufP + jOutOfs),
                                     &ckEncryptedLen);
 
-    ckAssertReturnValueOK(env, rv);
-
-cleanup:
     if (directIn == 0 && inBufP != NULL) {
```
Comment (valeriepeng) [inline src/jdk.crypto.cryptoki/share/native/libj2pkcs11/p11_crypt.c:172]: with this change, inBufP and outBufP should be non-null when directIn ==0 and directOut == 0, right?

Comment (valeriepeng) [reply src/jdk.crypto.cryptoki/share/native/libj2pkcs11/p11_crypt.c:172]: Same goes for the rest of the file. Otherwise, changes look good.

Comment (djelinski) [reply src/jdk.crypto.cryptoki/share/native/libj2pkcs11/p11_crypt.c:172]: Good point, will remove the extra checks.

**Comment by djelinski** [comment]
Thanks for the reviews!

/integrate

**Pull request closed without merging**